### ISSUE TITLE
Events / calendar

### ISSUE THREAD

**Issue by AyaMorisawa** [misskey-dev/misskey#3217]
> Invite (local?) users to events and filter / show events the current user is invited or public available events.

_Originally posted by @pwFoo in https://github.com/syuilo/misskey/issues/3110#issue-377169054_

**Comment by ssmucny** [comment]
A proposed implementation for events. Calendar view would be a separate issue.

## Compose Event
Add option in note compose modal to create an event note (similar to creating a poll). Note expands with _optional_ form fields.
- Title (string)
- Location (string)
- Start (datetime)
- End (datetime)
- Timezone (string from predefined list (ex: UTC−05:00))
- Invite List (array of user mentions)

The normal note text area will be below and contains the event "Details"

### Note Generation
When the note is posted it will be converted to a text representation using almost YAML format for the metadata to make it readable by remote instances that don't format events. It would look like this:
```YAML
Title: Event Title
Location: Event Location
Start: mm/dd/yyyy hh:mm (Time optional for start)
End: mm/dd/yyyy hh:mm (Time optional for end)
---
This event's details...

#event
Invite List: @user1 @user2 @user3@remote.instance
```
- In the note, all YAML fields above `---` are optional. If the value is empty then the field is not included in the note. If all fields are empty then `---` isn't included either.
- an `#event` tag is appended to end of note automatically to help with searching and recognition by parsers.
- Everything after the `---` is just normal note content and unstructured. This could alternatively be accomplished by placing it within a content warning fold.
- If Invites is not empty then the note is automatically posted as a direct note. Otherwise the user can set the visibility themselves. Invite list is not in YAML because `@` at beginning of words is not valid YAML and a long invite list would clutter the start of the event note.

## Viewing Event Note
When an event note is displayed on the timeline, it is detected by the `#event` tag included in the note. This will also pick up notes that aren't formatted as events, but those will just be detected as having only event details. The event will be formatted somewhat nicer like the below (empty fields excluded). The formatting does not change the note source text, just how it is displayed.

**Event Title**
📍 Event Location
🕙 Start - End
👥 Invites (on hover/click shows popup of list of invitees)

Event Details...
#event

### Responding
Below an event, The UI will display a row of buttons for Going, Maybe, and Decline. When one is clicked, it will react to the note with ✔️ , ❔, or ❌ respectively.  If a user clicks on another one of the buttons, the old reaction should be replaced with the newer one. Below the buttons, there should be a message for "n going, n maybe, n declined" with each one being clickable/hover to show people who reacted with that. The reserved reaction emoji should not be displayed next to the other normal emoji reactions.

## Notes on Federation
I think the federated behavior of this proposal is reasonable. Events federated to other instances and remote users can even respond to events (though if they don't use compatible software then they may not know what reserved reactions to use or be unable to use unicode reactions). Users from the local instance also will be able to respond to remote events as if they were local, though they may not be formatted as nicely.

## Future Calendar Feature
A future calendar feature could filter notes on the `#event` tag and those with user as invitee, or where user reacted. These could then be displayed in a filterable/sortable list/agenda/calendar view.

## Required Code Changes
Because of the design, this change would be frontend only and require no backend changes. This is by design to integrate more easily with existing infrastructure and other fediverse software without event support.

**Comment by ssmucny** [comment]
I would be willing to start implementing this feature. I see this as broken down into the following components (roughly in order of development):
- Add database support to attach an event to a note
- Add/modify API endpoints to manage event / include event in notes
- Add UI elements to create and display an event in a note
- Federation of events with other services (Mobilize, Friendica, etc.) / export events in iCal format
- support for users to respond to events (RSVP)
- Dedicated views/API for events
  - calendar
  - event query/search/list

I can break this out into separate issues and provide HLD/LLD for the components or include them here. I am in conversation with some companies that are interested in this that may be able to provide funding as well (waiting on RFP).

**Comment by ssmucny** [comment]
# Database Events
- Add new table for events
- foreign key 1-to-1 with notes (each event has 1 parent note, a note may have an event). An event will be considered a type of note that has extra properties.
- contains (in addition to normal note properties) start/end datetimes (end is optional), title, metadata
- index on start/end
- metadata is jsonb field with key/value list of all optional fields (just text values) to enable ad hoc filters/searches
  - location
  - neighborhood
  - transit
  - virtual
  - cost
  - etc.
- If the parent note is deleted, then the event should be deleted

**Comment by ssmucny** [comment]
# Events API
- Modify notes API to return an event member in the note object if present (similar to renotes)
- events, because they are a type of note, will have the same ACL as notes (direct, followers, home, public, local)
- Add new API endpoint to search for events with following options (options should be included as query/URL parameters to enable permalinks to searches)
  - list of users to use as source (like antennas) (probably just user IDs, but maybe support usernames if starts with @[%40 encoded])
  - start date and/or end date or start and/or end noteId (default to today as start date). Allow several keywords like `today` and `tomorrow` in addition to ISO dates
  - limit (default 10)
  - sort by (created at or start time). default to start time. If created at then sort newest first, if start time sort oldest/soonest first.
  - metadata filters (mapping of `string -> [string]`) where each list of strings is an OR for the key in the metadata object. a value of (BLANK) allows NULL to match. The keys are ANDed together. keys that don't match other parameters are treated as metadata filters.
  - ex: `https://thecle.land/api/events/search?user=abc123+ef347&start=2023-03-31&neighborhood=Downtown+Flats+Ohio%20City&cost=Free+(Blank)` will find events posted by user IDs `abc123` and `ef347`, that start on or after `2023-03-31`, and have metadata fields: neighborhood that is one of `Downtown`, `Flats`, or `Ohio City`, and cost either not listed or `Free`.

**Comment by acid-chicken** [comment]
Thank you for your work!

> - foreign key 1-to-1 with notes (each event has 1 parent note, a note may have an event). An event will be considered a type of note that has extra properties.

Is the reason for this design because the notebook is responsible for outlining and describing the event?

**Comment by ssmucny** [comment]
> Thank you for your work!
> 
> > * foreign key 1-to-1 with notes (each event has 1 parent note, a note may have an event). An event will be considered a type of note that has extra properties.
> 
> Is the reason for this design because the notebook is responsible for outlining and describing the event?

Yes. The note text could be considered the description of the event. Also any attachments to the note would be inherited as well by the event. The event just adds new properties that are event specific. An alternative would be to add new columns to the note table for events, but they would be empty most of the time. Separating the event specific properties into a separate table should improve clarity.

**Comment by acid-chicken** [comment]
Then, maybe #8362 should be accomplished at the same time.

**Comment by ssmucny** [comment]
I have opened a draft PR to track progress and discuss implementation at https://github.com/misskey-dev/misskey/pull/10628
